service-base/src/main/java/org/eclipse/hono/service/metric/MicrometerBasedMetrics.java

```diff
@@ -81,6 +82,18 @@
      * The name of the meter for command messages.
      */
     public static final String METER_COMMANDS_RECEIVED = "hono.commands.received";
+    /**
+     * The name of the meter for queue full events.
+     */
+    public static final String METER_DOWNSTREAM_FULL = "hono.downstream.full";
+    /**
+     * The name of the meter for sent messages.
+     */
+    public static final String METER_DOWNSTREAM_SENT = "hono.downstream.sent";
+    /**
+     * The name of the meter for timed out messages.
+     */
+    public static final String METER_DOWNSTREAM_TIMEOUT = "hono.downstream.timeout";
```

Comment:
for the other metrics we followed an approach of using a tag to indicate the outcome of the operation. Is there a particular reason why we wouldn't want to follow that approach here as well?

Reply:
Yes.  I put a note into the `timeout()` method, maybe there is a better place:

~~~java
/*
 * We report timeouts with a different meter, since the message might still be
 * accepted by the remote peer, at a time after the timeout expired. And so we
 * can still track those times.
 */
~~~

I an nutshell, timeout and "completion" are actually two different events. For timeout we simply decide to, no longer listen to the outcome of the "send message" operation. Still we should get back a disposition, and thus would have the total run time of the operation, which we can log. This would give us an understanding of how long operations actually run.

For the "queue full" situation we don't even start sending. So it doesn't seem to make sense to capture any time.